app/src/main/java/de/westnordost/streetcomplete/data/osm/osmquest/OsmFilterQuestType.kt

```diff
@@ -16,11 +15,11 @@ abstract class OsmFilterQuestType<T> : OsmElementQuestType<T> {
         /* this is a considerate performance improvement over just iterating over the whole MapData
         *  because for quests that only filter for one (or two) element types, any filter checks
         *  are completely avoided */
-        val iterable = MultiIterable<Element>()
-        if (filter.includesElementType(Element.Type.NODE)) iterable.add(mapData.nodes)
-        if (filter.includesElementType(Element.Type.WAY)) iterable.add(mapData.ways)
-        if (filter.includesElementType(Element.Type.RELATION)) iterable.add(mapData.relations)
-        return iterable.filter { element -> filter.matches(element) }
+        var elements = sequenceOf<Element>()
+        if (filter.includesElementType(Element.Type.NODE)) elements += mapData.nodes
+        if (filter.includesElementType(Element.Type.WAY)) elements += mapData.ways
+        if (filter.includesElementType(Element.Type.RELATION)) elements += mapData.relations
+        return elements.filter { element -> filter.matches(element) }.asIterable()
```

Comment:
I am personally allergic to if statements without braces (except when replacing the ternary operator), and using them on one liners like this makes them very busy. Here's a rewrite to avoid that, and use a more functional style:

```kotlin
        val elements = mapOf(
            Element.Type.NODE to mapData.nodes,
            Element.Type.WAY to mapData.ways,
            Element.Type.RELATION to mapData.relations
        ).filterKeys(filter::includesElementType).values.flatten()
        return elements.asSequence().filter(filter::matches).asIterable()
```

Reply:
If you don't like using the function references (they work by reflection, and do smell a little bit of magic):

```kotlin
        ).filterKeys { filter.includesElementType(it) }.values.flatten()
        return elements.asSequence().filter{ filter.matches(it) }.asIterable()
```

Reply:
Alternative version using `sequence()` (maybe better performance?)
```kotlin
        return sequence {
            mapOf(
                Element.Type.NODE to mapData.nodes,
                Element.Type.WAY to mapData.ways,
                Element.Type.RELATION to mapData.relations
            ).filterKeys(filter::includesElementType).forEach { yieldAll(it.value) }
        }.filter(filter::matches).asIterable()
```

Reply:
Whoa, really? Readability aside, you:
1. `mapOf`: create a hashmap and add three values
2. `filterKeys`: you create yet another new hashmap with only the keys added that return true on `filter::includesElementType`
3. `.values`: you access the value set of the hashmap (Not sure if this requires extra processing. It would if the set value set is created lazily or if it is not a view onto the hashmap but a copy of all values in it)
4. `flatten`: you create a list of these (all the values in the map) by iterating through it all and adding it all to an array list
5. since in step 4, you already copied the complete lists, the last line doesn't really make sense

That's a lot of transforming and copying of data just to avoid few `if`s.

Your last suggestion would at least work as far as I understand because you don't flatten everything into a list.

Reply:
On the `sequence { }` syntax, I think it would look like this then. Since anyone who reads this also needs to understand sequences, this would be fine too. Not sure if `flatten` keeps the sequence a sequence, it should otherwise the below code doesn't make sense.


```kotlin
return sequence {
  if (filter.includesElementType(Element.Type.NODE)) yield(mapData.nodes)
  if (filter.includesElementType(Element.Type.WAY)) yield(mapData.ways)
  if (filter.includesElementType(Element.Type.RELATION)) yield(mapData.relations)
}.flatten().filter(filter::matches).asIterable()
```

Reply:
> Not sure if `flatten` keeps the sequence a sequence

It does: https://github.com/JetBrains/kotlin/blob/8fa848bed384568a8ceb9731ec9d45663aa64521/libraries/stdlib/src/kotlin/collections/Sequences.kt#L94-L99

What about `yieldAll` though?

```kt
        return sequence {
            if (filter.includesElementType(Element.Type.NODE)) yieldAll(mapData.nodes)
            if (filter.includesElementType(Element.Type.WAY)) yieldAll(mapData.ways)
            if (filter.includesElementType(Element.Type.RELATION)) yieldAll(mapData.relations)
        }.filter { filter.matches(it) }.asIterable()
```

Reply:
Right, that should work too. (Disclaimer: I am not an expert on Kotlin sequences)

Reply:
Well, I *did* phrase it as a question (***maybe*** better performance<!---->***?*** ). Disclaimer: I am also not an expert on Kotlin sequences, although I re-read most of the documentation before writing my version, so some details are fresh in my mind.

4\) This is *mostly* what I was talking about — that my second version was more efficient than the first version, because it avoids `flatten()` on a list, which will evaluate eagerly.

5\) Objects are copied by reference in java, so while there is a performance impact, it may be smaller than it seems. Thus, a lazy filter may still provide a meaningful performance impact. It depends on the cost of running `filter.matches`.

1-3) Because of the small number of elements (3 max), this is essentially a fixed cost. I don't know exactly how big that cost (2x hash map creation, up to 6 hash map insertions, up to 3 hash map accesses) is. I did not consider it in my performance analysis, but it may have a meaningful impact.

However, the main reason I say "maybe" is that I don't know the implementation of `elements += mapData.nodes`. In particular, **I do not know if it is eagerly *copying* the nodes into an array inside the `elements` sequence (which would then be evaluated lazily), or just holding a reference to `mapData.nodes`. I have the same question about `yieldAll()`, although I would guess that it is *more likely* to be optimized in this way. This is the key answer needed to analyze the efficiency.

The block that you pass to `sequence()` is a suspend function ([docs here](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.sequences/sequence.html)). Each time you ask for values in the sequence, the block runs only up until you yield enough values, then suspends. There is some cost to suspending (should be low, since coroutines are lightweight, but I don't know *how* lightweight), and a memory cost to capture the closure.

---

So this is why I am uncertain. There are many different unknown variables.

I would suspect that the fastest / approach is either the latest that @FloEdelmann wrote, with `if`s and `yieldAll()`, or this:

```kotlin
        return sequenceOf(
            Pair(Element.Type.NODE, mapData.nodes),
            Pair(Element.Type.WAY, mapData.ways),
            Pair(Element.Type.RELATION, mapData.relations)
        )
            .filter { (type, _) -> filter.includesElementType(type) }
            .map { (_, elements) -> elements }.flatten()
            .filter { filter.matches(it) }
            .asIterable()
```